### PR TITLE
8308593: Add Keepalive Extended Socket Options Support for Windows

### PR DESCRIPTION
**Pull request by tkyc**
The PR adds support for the keepalive extended socket options on Windows. For TCP_KEEPIDLE and TCP_KEEPINTVL, these options are supported starting from Windows 10 version 1709. TCP_KEEPCNT is supported starting from Windows 10 version 1703. Information on these socket options can be found [here](https://learn.microsoft.com/en-us/windows/win32/winsock/ipproto-tcp-socket-options).

I've also corrected the `handleError()` function. On Windows, the error needs to be retrieved using `WSAGetLastError()` and error codes are prefixed with "WSA". Information on this can be found [here](https://learn.microsoft.com/en-us/windows/win32/winsock/error-codes-errno-h-errno-and-wsagetlasterror-2).

>The error codes returned by Windows Sockets are similar to UNIX socket error code constants, but the constants are all prefixed with WSA.

>Error codes set by Windows Sockets are not made available through the errno variable.

No new tests were added as the existing tests should cover this.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8308593](https://bugs.openjdk.org/browse/JDK-8308593): Add KEEPALIVE Extended Socket Options Support for Windows (**Enhancement** - P4)

### Reviewers
 * [Michael McMahon](https://openjdk.org/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**) ⚠️ Review applies to [2ada2207](https://git.openjdk.org/jdk/pull/14232/files/2ada2207c1f9fa16da3b1ca2977229403c712852)
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - **Reviewer**)
 * [Vyom Tewari](https://openjdk.org/census#vtewari) (@vyommani - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14232/head:pull/14232` \
`$ git checkout pull/14232`

Update a local copy of the PR: \
`$ git checkout pull/14232` \
`$ git pull https://git.openjdk.org/jdk.git pull/14232/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14232`

View PR using the GUI difftool: \
`$ git pr show -t 14232`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14232.diff">https://git.openjdk.org/jdk/pull/14232.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14232#issuecomment-1570719929)

### CODE REVIEW

**Comment by AlanBateman** [comment]
I think you may have missed my comment in the JBS issue where I asked about the semantics of SIO_KEEPALIVE_VALS and whether we could implement TCP_KEEPIDLE and TCP_KEEPINTERVAL on Windows by knowing what the defaults are. The proposal that you have here is a new Windows specific and JDK-specific socket option that requires a lot of thinking about before going that route.

**Comment by AlanBateman** [comment]
In the discussion on net-dev/nio-dev, Daniel Jeliński pointed out that Windows does have TCP_KEEPIDLE and TCP_KEEPINTVL since Windows 10 version 1709. So I assume this PR can be returned to closed or re-returned to draft while the implementation changes are developed to make use of these socket options.

**Comment by tkyc** [comment]
HI, just wondering if there's anything needed from me further (the related threads have been inactive for some time).

**Comment by AlanBateman** [comment]
> HI, just wondering if there's anything needed from me further (the related threads have been inactive for some time).

The original to use TCP_SIO_KEEPALIVE was problematic for several reasons, changing it to use TCP_KEEPCNT, TCP_KEEPIDLE, and TCP_KEEPINTVL after finding out that these socket options are supported on Windows is good. So I think we should review and include this.

**Comment by jaikiran** [comment]
The changes to `WindowsSocketOptions.java` which are merely wiring changes, look OK to me. I don't have experience with Windows nor much JNI code, so I'll let other reviewers take a look.

I see that the JNI code is updated to use `WSAGetLastError()` instead of `errno` to get the error code. You already provided a link to the documentation of that API and reading through it, it appears to be the right thing to do and it seems to be supported on all relevant Windows versions of interest. However, that change does impact other existing code in that file. Do you think, we could do that change in a separate PR?

In the meantime, could you please update the copyright year on both these changed files to `2022, 2023, ` from the current `2022, `?

**Comment by jaikiran** [comment]
Hello Terry,

> No new tests were added as the existing tests should cover this.

The existing `test/jdk/java/net/SocketOption/TcpKeepAliveTest.java` test should indeed cover these changes. It checks the `supportedOptions()` and if it finds these `TCP_KEEPCOUNT`, `TCP_KEEPIDLE` and `TCP_KEEPINTERVAL` to be supported then it runs the relevant tests. I'll run some tests with the current state of this PR to make sure these tests are running as expected.

**Comment by Michael-Mc-Mahon** [comment]
Webrev 01 seems to be missing a number of files.

**Comment by tkyc** [comment]
Thanks all for the review so far. I'll make the copyright year changes. So, the consensus for the `errno` to `WSAGetLastError` change is that it's the right thing to do. But, I should make this change in another PR? Are we all in agreement with that?

**Comment by jaikiran** [comment]
I think Michael means that you can keep the `WSAGetLastError` change in this current PR. I don't have any objection to that.

**Comment by Michael-Mc-Mahon** [comment]
> I think Michael means that you can keep the `WSAGetLastError` change in this current PR. I don't have any objection to that.

Right. That's what I meant. The change looks good to me, but I might run the patch through the test system here to see how it works with the existing tests in this area and the change above doesn't cause any issues.

**Comment by AlanBateman** [comment]
> I think it is right to change the `errno` to `WSAGetLastError`. There is an errno.h in VC++ which must be getting included indirectly. It is used in hotspot. But, we shouldn't be using it in the networking code.

Right, WSAGetLastError should be used with Windows Sockets, not errno.

**Review by Michael-Mc-Mahon — APPROVED** [review]
Looks good to me, subject to copyright update.

**Comment by tkyc** [comment]
Awesome, thanks all for the review.

**Review by djelinski — APPROVED** [review]
LGTM. Verified that `test/jdk/java/net/SocketOption/TcpKeepAliveTest.java` is able to set and get the options on Windows 11 as expected.

On a side note, the test could be more verbose - currently it is not clear if the test passes because the options worked or because they were unsupported. But that's preexisting and not much of a problem anyway.

File: src/jdk.net/windows/classes/jdk/net/WindowsSocketOptions.java
Comment (vyommani) [inline src/jdk.net/windows/classes/jdk/net/WindowsSocketOptions.java:93]: there is inconsistency in function naming some place you use 'keep' and other places you use 'Keep', can you please make it consistence ?.

Comment (tkyc) [reply src/jdk.net/windows/classes/jdk/net/WindowsSocketOptions.java]: Fixed inconsistency.

**Review by vyommani — APPROVED** [review]
thanks for doing changes for other platforms as well. Latest changes looks OK to me.

**Comment by Michael-Mc-Mahon** [comment]
It looks like you need to type the "integrate" command again, after the last change. We can sponsor it then.

**Comment by jaikiran** [comment]
> On a side note, the test could be more verbose - currently it is not clear if the test passes because the options worked or because they were unsupported. But that's preexisting and not much of a problem anyway.

Agreed. When I tested this PR in our CI, I had to locally update `TcpKeepAliveTest.java` to make sure the options were indeed being considered. I'll open a separate JBS issue and propose a PR with these log messages changes in that test.

**Review by djelinski — APPROVED** [review]

**Comment by tkyc** [comment]
Thanks all. /integrate

**Comment by dfuch** [comment]
Hi, this issue is marked as needing a release note but none has been produced.
Could you please add a release note to [JDK-8308593](https://bugs.openjdk.org/browse/JDK-8308593) for this?
Instructions are available here: https://openjdk.org/guide/#release-notes

**Comment by tkyc** [comment]
@dfuch Sorry I missed that step. I'll add a release note according to your linked instructions.

**Comment by tkyc** [comment]
Just a heads up, incase things move slow. Since I don't have a JBS account, I'm just waiting on some help from my msft liaison to help add the release note to the ticket.

**Comment by tkyc** [comment]
Hello, a release note was added to the ticket awhile ago. Let me know if there's anything else I need to do or correct. Thanks again.

**Comment by dfuch** [comment]
Thanks Terry. I have made some updates and marked it DELIVERED. Can you double check that everything still looks good?

**Comment by tkyc** [comment]
@dfuch LGTM. Thanks for the quick update Daniel.

**Pull request closed without merging**
